Paginate exports

orangejenny opened the pull request:
<img width="1466" alt="screen shot 2019-01-21 at 5 32 08 pm" src="https://user-images.githubusercontent.com/1486591/51502106-87194b00-1da2-11e9-8c5b-281f015a1151.png">

@esoergel this turned out bigger than I expected, but the commit history is fairly logical.

Marking don't merge because this shouldn't be deployed until https://github.com/dimagi/commcare-hq/pull/23060 is deployed on all environments

fyi @dimagi/product 

Jira: https://dimagi-dev.atlassian.net/browse/PF-257

orangejenny committed:
Updated pagination to fetch all doc values, then a page worth of full docs

orangejenny committed:
Updated exports list to fetch exports via ajax.

- Changed js arrays of exports to observables
- Added knockout logic to handle spinner and error
- The new ajax view is just returning a blank list for now

orangejenny committed:
Updated get_exports_list to fetch data instead of returning an empty list.

This extracts all of the export-getting code in the *ListView classes
into a parallel set of *ListHelper classes.

orangejenny committed:
Added _priv_check logic to new helper classes

orangejenny committed:
Removed HQJSONResponseMixin

This is an angular mixin, no longer needed.

orangejenny committed:
Moved most logic remaining in *ListView classes to *ListHelper classes.

This leaves the *ListView classes doing nothing but providing page_content,
page_title, etc (the properties & functions inherited from BaseDomainView).
That seems like a cleaner split than having some logic in the *ListViews,
some logic in *ListHelpers, and some logic duplicated (like the form_or_case
and is_deid properties).

orangejenny committed:
Extracted exportPanelModel

Needed so that the two panels can be paginated independently.

orangejenny committed:
Added pagination

orangejenny committed:
Moved can_view from model into view logic

orangejenny committed:
Replaced is_safe with is_deidentified

The is_safe property being used previously just checks the value
of 'is_deidentified', which is available in map.js.

orangejenny committed:
Updated bulk export UI.

Added a count of exports so it's more apparent you're only getting exports from the first page.
It isn't possible to select exports for bulk export across pages, but you
can make each page show up to 100 exports.

Also moved the bulk export notice to the top of the page instead of showing it in each panel.
The multi-panel UI that you get when the EXPORT_OWNERSHIP flag is on is a bit
weird when combined with the bulk export functionality (why are the "All"/"None" buttons
inside of the panels, but they affect both panels?) and could use some UI attention.

orangejenny committed:
Renamed get_exports_list tp get_exports_page

djmore9 commented:
❤️❤️❤️❤️ Thank you!

orangejenny committed:
Fixed lint error

esoergel commented on corehq/apps/export/static/export/js/export_list.js:
```diff
@@ -158,12 +158,14 @@ hqDefine("export/js/export_list", [
 
     var exportListModel = function (options) {
-        assertProperties.assert(options, ['isDeid', 'modelType', 'urls']);
+        assertProperties.assert(options, ['isDailySavedExport', 'isDeid', 'isFeed', 'modelType', 'urls']);
```
Is this same JS model used for both Daily Saved Exports and regular ones?  Are they the same pages?  The appearance and functionality is pretty different.  Certainly makes sense to share pagination logic however, just surprised that this is used for both.

orangejenny replied:
Yeah, the daily saved / dashboard feeds UIs are supersets of the basic export page: same basic structure, but daily saved adds the update data button & progress bar to the first column and also additional columns for enable/disable and editing filters.

I think it might be an improvement to redesign the daily saved & dashboard feed pages so that they're *not* just extensions of base exports, because there's just so much crammed into that first column (progress bar, etc) but that's a bigger question for the future.

esoergel commented on corehq/apps/export/views/list.py:
```diff
@@ -56,52 +56,52 @@
-
-        return super(BaseExportListView, self).dispatch(request, *args, **kwargs)
+    @classmethod
+    def get(self, request, domain, form_or_case=None, is_daily_saved_export=False, is_feed=False, is_deid=False):
```
You could cut down on some of these subclasses by passing `is_deid` to the constructor and setting it as a property.  The functionality doesn't differ that much.

esoergel commented on corehq/apps/export/views/list.py:
```diff
@@ -175,17 +174,252 @@ def _fmt_emailed_export_fileData(self, fileId, size, last_updated,
+            "case": EditCaseDailySavedExportView
+        }[model]
+
+    def fmt_export_data(self, export):
```
Guessing these methods are too dissimilar to cleanly deduplicate?

orangejenny replied:
yeah, going to leave these methods as they are

orangejenny replied:
you were right about this - cleaned them up in https://github.com/dimagi/commcare-hq/pull/23071

esoergel commented on corehq/apps/export/views/list.py:
```diff
@@ -56,52 +56,52 @@
+                return DeIdFormExportListHelper(request, domain)
+            return FormExportListHelper(request, domain)
+        elif form_or_case == 'case':
+            return CaseExportListHelper(request, domain)
```
You could also set the appropriate helper class as a property of each view class and avoid needing to do this dispatch method (unless of course your plan is to eventually deduplicate those view classes).

orangejenny replied:
Not quite following - the only place I call this is in the `get_exports_page` functional view. How does having the helper be a property of the class-based list views help?

esoergel replied:
Oh yeah, nevermind - I think my comment made more sense when the CBVs were using this method, but since there's only one functional view, this approach works better.

esoergel commented on corehq/apps/export/views/list.py:
```diff
@@ -91,7 +91,13 @@ def __init__(self, request, domain):
         self.permissions = ExportsPermissionsManager(self.form_or_case, domain, request.couch_user)
 
+    def _priv_check(self):
+        return True
```
This helper class seems like an odd place for a permissions check - I'd expect it to happen in the view function.

orangejenny replied:
Yeah, I debated what to do with this logic, ultimately moved it here for the sake of DRYer code - if it happens in the view function I need to duplicate the logic in the `get_exports_page` view.

esoergel commented on corehq/apps/export/views/list.py:
```diff
@@ -43,58 +42,62 @@
-
-        return super(BaseExportListView, self).dispatch(request, *args, **kwargs)
+    @classmethod
+    def get(self, request, domain, form_or_case=None, is_daily_saved_export=False, is_feed=False, is_deid=False):
```
It looks like the helper classes only needs `request` for `domain` and `couch_user` - you could drop it and pass those in directly, which would make this even more of a standalone helper.

orangejenny replied:
good point, I'll look at updating this code

orangejenny replied:
This isn't true (might have changed in a later commit): the helpers need access to `request.build_absolute_uri` and `is_location_safe_for_user` also needs to the request. I did add a commit to remove `domain` from the param list since it's already available.

esoergel commented on corehq/apps/export/views/list.py:
```diff
@@ -517,20 +555,15 @@ def update_emailed_export_data(request, domain):
 
 @location_safe
-class DailySavedExportListView(BaseExportListView):
+class DailySavedExportListView(BaseExportListView, DailySavedExportListHelper):
```
Did you look into accessing the helper functionality by composition rather than inheritance?  It's strange for these to be used both as a view mixin and as a standalone utility class.

orangejenny replied:
Agreed this is unusual, I think I did inheritance because the boundary between what `self` does and what the helper does is a bit blurry (like as you mentioned above, the privilege check is in the helper but one might expect it to be in the view).

I'd like to change all of the class-based views in this file to functional views and have virtually all of the logic contained in the helper, but I ran out of energy. I'll take another look at doing that, I think that'd address a bunch of the slightly weird things about this PR.

orangejenny replied:
Went back and looked at changing all the class-based views to functional ones and it's more involved than I remembered - it means making a bunch of updates to the menu code (not surprising) and to other views (which use them to create links between pages within exports). I'm going to call this out of scope right now.

esoergel commented on corehq/apps/export/dbaccessors.py:
```diff
@@ -66,25 +66,42 @@ def get_form_inferred_schema(domain, app_id, xmlns):
     return _get_export_instance(CaseExportInstance, key)
 
 
+def get_brief_form_export_instances(domain):
```
I know we often use multiple dbaccessor functions rather than use a series of parameters to control the return values, but in this case I think the latter would simplify things, since there are so many permutations in use.

orangejenny replied:
will update

orangejenny replied:
17ab783e74c - I did keep one of the methods split into a `get` and `get_brief` version, because it got awkward dealing with objects and dicts in the same method.

esoergel commented on corehq/apps/export/static/export/js/export_list.js:
```diff
@@ -1,8 +1,22 @@
+ *  There are three models in this file:
+ *      exportModel represents an individual export: name, description, case type, etc.
+ *      exportPanelModel represents a set of exports. Each of these is displayed in the UI as
+ *          a panel with independent pagination.
```
nit: "page" is a more familiar term to me than "panel"

orangejenny replied:
I don't understand - how would you phrase this?

esoergel replied:
`exportPageModel` or something.  When I read `exportPanelModel`, I thought it was a panel representing an individual export, rather than a set of exports.

esoergel commented on corehq/apps/export/templates/export/partials/export_list_controller.html:
```diff
@@ -41,15 +41,11 @@
-                <div data-bind="visible: exports().length > 6">
-                    {% include 'export/partials/export_bulk_notice.html' %}
-                </div>
+                <pagination data-apply-bindings="false"
```
That's nifty

esoergel commented on corehq/apps/export/models/new.py:
```diff
@@ -809,9 +809,6 @@ def generate_instance_from_schema(cls, schema, saved_export=None, auto_select=Tr
         return instance
 
-    def can_view(self, user_id):
-        return self.owner_id is None or self.sharing != SharingOption.PRIVATE or self.owner_id == user_id
```
:+1:

esoergel commented on corehq/apps/export/views/list.py:
```diff
@@ -144,7 +144,7 @@ def _can_view(e, user_id):
                 and ('owner_id' in export and export['owner_id'] == self.request.couch_user.user_id) == my_exports
             ]
         if self.is_deid:
-            brief_exports = [x for x in brief_exports if x['is_safe']]  # TODO: is_deidentified
```
:heart:

orangejenny replied:
FWIW, comments imply that `is_safe` can maybe be killed altogether: https://github.com/dimagi/commcare-hq/blob/dc382515afc266efd82a0237d011849e45e3ff76/corehq/apps/export/models/new.py#L707

I looked briefly at removing it altogether but wasn't sure what to do with code like this - https://github.com/dimagi/commcare-hq/blob/dc382515afc266efd82a0237d011849e45e3ff76/corehq/apps/export/custom_export_helpers.py#L132 - which implies that some exports do have an `is_safe` property...but maybe those are only old exports?

esoergel reviewed (approved):
Looks great!

You don't need to wait for a deploy of the view changes to merge, unless you wanted to do a preindex separately from the deploy (which I don't think is necessary for the meta db).  The switchover to the new code won't happen until the preindex is finished.

I left a lot of comments pontificating from my ivory tower without being in the weeds doing these refactors.  Nothing is blocking, feel free to address any that resonate with you and ignore the rest.

orangejenny commented on corehq/apps/export/templates/export/partials/export_list_controller.html:
```diff
@@ -41,7 +41,13 @@
+                TODO: this is solely so the bulk button appears at both the top and bottom of the screen if there
+                are a lot of exports. Paginating should eliminate the need for this.
+
+                Note this logic and this second div happens 3 times in this file.
```
fixed in later commit :)

orangejenny commented on corehq/apps/export/static/export/js/export_list_main.js:
```diff
@@ -34,11 +34,11 @@ hqDefine("export/js/export_list_main", [
 
         var modelType = initialPageData.get("model_type");
         $("#export-list").koApplyBindings(listModels.exportListModel({
-            exports: initialPageData.get("exports"),
```
ignoring this one for the sake of simpler code

orangejenny commented on corehq/apps/export/static/export/js/export_list.js:
```diff
@@ -190,6 +192,8 @@ hqDefine("export/js/export_list", [
                     is_deid: self.isDeid,
                     model_type: self.modelType,
+                    is_daily_saved_export: self.isDailySavedExport ? 1 : 0,
+                    is_feed: self.isFeed ? 1 : 0,
```
haha, yep, same discussion. I think the 0/1 logic was pre-existing in this case (or if i added it in this PR, i did so to be consistent with other 0/1 logic in this file)...I really don't think any of the ways to deal with true/false values are ideal.

orangejenny commented:
> The switchover to the new code won't happen until the preindex is finished.

Aha, thank you for the certainty - I have done this kind of thing wrong in the past.

orangejenny committed:
Removed domain parameter from ExportListHelper.get

orangejenny committed:
Replaced 'brief' dbaccessors with include_docs param

orangejenny commented:
@esoergel Thank you for the feedback (even the parts I didn't address - I'll keep them in mind for the future, suspect this is not the last time I'll be touching the exports UI). Made a couple of updates, planning to deploy tomorrow or Thursday.

orangejenny committed:
Updated default items per page to 25

orangejenny committed:
Replaced Copy with Delete on export list

orangejenny committed:
Moved columns around.

Moved the primary action (export) to the left (more promiment),
the dangerous action (delete) to the right (less prominent),
and moved the bulk export checkbox next to the export button.

orangejenny committed:
Updated bulk export column to be narrower (w=1)

orangejenny committed:
DRY up fmt_export_data

orangejenny committed:
Replaced Delete with Copy on export edit

orangejenny committed:
Merge pull request #23071 from dimagi/jls/exports-swap-buttons

Swap delete and copy buttons on exports

orangejenny committed:
Fixed lint error

esoergel reviewed (approved):
Looks good!

orangejenny commented on corehq/apps/export/views/list.py:
```diff
@@ -104,20 +107,90 @@ def bulk_download_url(self):
         """
-        raise NotImplementedError("must implement fmt_export_data")
+        from corehq.apps.export.views.new import DeleteNewCustomExportView
+        formname = export.formname if isinstance(export, FormExportInstance) else None
```
cool, good to know

biyeun commented:
:tada::tada::tada::tada::tada::tada::tada::tada:
